python/nlp/document.py

```diff
@@ -0,0 +1,239 @@
+
+  @length.setter
+  def length(self, value):
+    self.frame[self.names.phrase_length] = value
+
+  @property
+  def end(self):
+    return self.begin + self.length
+
+  def evokes(self):
+    return self.frame(self.names.phrase_evokes)
+
+  def evoke(self, evoked):
+    return self.frame.append(self.names.phrase_evokes, evoked)
+
+
+class Document(object):
+  def __init__(self, frame=None, store=None, names=None):
+    # Create store, frame, and names if missing.
+    if frame != None:
+      store = frame.store()
+    if store == None:
+      store = sling.Store()
+    if names == None:
+      names = DocumentNames(store)
+    if frame == None:
+      frame = store.frame([(names.isa, names.document)])
+
+    # Initialize document from frame.
+    self.frame = frame
+    self.names = names
+    self.tokens = []
+    self.mentions = []
+    self.themes = []
+    self.tokens_dirty = False
+    self.mentions_dirty = False
+    self.themes_dirty = False
+
+    # Get tokens.
+    tokens = frame[names.document_tokens]
+    if tokens != None:
+      for t in tokens:
+        token = Token(names, t)
+        self.tokens.append(token)
+
+    # Get mentions.
+    for m in frame(names.document_mention):
+      mention = Mention(names, m)
+      self.mentions.append(mention)
+
+    # Get themes.
+    for theme in frame(names.document_theme):
+      self.themes.append(theme)
+
+  def add_token(self, text=None, start=None, length=None, brk=SPACE_BREAK):
+    slots = [
+      (self.names.isa, self.names.token),
+      (self.names.token_index, len(self.tokens)),
+    ]
+    if text != None: slots.append((self.names.token_text, text))
+    if start != None: slots.append((self.names.token_start, start))
+    if length != None: slots.append((self.names.token_length, length))
+    if brk != SPACE_BREAK: slots.append((self.names.token_break, brk))
+    token = Token(self.names, self.frame.store().frame(slots))
+    self.tokens.append(token)
+    self.tokens_dirty = True
+    return token;
+
+  def add_mention(self, begin, end):
+    length = end - begin
+    slots = [
+      (self.names.isa, self.names.phrase),
+      (self.names.phrase_begin, begin),
+    ]
+    if length != 1: slots.append((self.names.phrase_length, length))
+    mention = Mention(self.names, self.frame.store().frame(slots))
+    self.mentions.append(mention)
+    self.mentions_dirty = True
+    return mention;
+
+  def add_theme(self, theme):
+    self.themes.append(theme)
+    self.themes_dirty = True
+
+  def update(self):
+    # Update tokens in document frame.
+    if self.tokens_dirty:
+      array = []
+      for token in self.tokens: array.append(token.frame)
+      self.frame[self.names.document_tokens] = array
+      self.tokens_dirty = False
+
+    # Update mentions in document frame.
+    if self.mentions_dirty:
+      slots = []
+      for mention in self.mentions:
+        slots.append((self.names.document_mention, mention.frame))
+      del self.frame[self.names.document_mention]
```

Comment:
this only gets rid of the first mention right?
did you mean to do:
self.frame[self.names.document_mention] = None

Similarly below for theme.

Reply:
The del frame[role] construction calls PyFrame::Assign() which uses Store::Delete(). The Store::Delete() method deletes all slots with a particular name, so I think this is okay.